### DISCUSSION TITLE
Incorrect Content-type returned

### DISCUSSION THREAD

**Discussion by ghost** [datopian/giftless#79] [Q&A] [answered]
Hi!

Not quite sure on which side this should be fixed...

CKAN: 2.9.1
ckanext-blob-storage: 0.7.0
datapusher: 0.0.16
giftless: 0.3.0
GCP for file storage

The issue is whenever a csv-file is added to datasets, it creates a request for datapusher to download it from CKAN. CKAN sends HTTP 302 with the redirection URL pointing to Giftless (http://127.0.0.1:9419/ in the screenshot). Then Giftless responses with the file content but sends the wrong Content-type, text/html instead of text/csv, this in turn causes a crash inside datapusher:

`Traceback (most recent call last):
  File "python3.8/site-packages/apscheduler/scheduler.py", line 512, in _run_job
    retval = job.func(*job.args, **job.kwargs)
  File "datapusher/jobs.py", line 439, in push_to_datastore
    row_set = get_row_set(table_set)
  File "datapusher/datapusher/jobs.py", line 438, in <lambda>
    lambda table_set: table_set.tables.pop())
IndexError: pop from empty list`

If content-type is set in ad-hoc manner to the correct one i.e. text/csv, then datapusher works nicely. But I guess it is probably Giftless responsibility to give the correct document type as content-type (?)

And the ultimate resulting problem is that csv-files are not previewed in CKAN UI.

![Request1](https://user-images.githubusercontent.com/79199982/110100535-31019900-7dab-11eb-8153-2e9c23e305b2.png)
![Request2](https://user-images.githubusercontent.com/79199982/110100594-3f4fb500-7dab-11eb-855c-87c028019707.png)

Any help in resolving this is highly appreciated!

**Comment by shevron** [comment, marked as answer]
What storage backend are you using (assuming local / streaming based on the info)? Not all backends support properly sending content-type headers. 

To give some background, Content-type is not somethnig handled by the Git LFS protocol (typically this isn't needed by Git LFS). Some of the storage backends supported by Giftless have extra functionality to handle content-type headers, but local storage doesn't store information about the content type anywhere, and so adding support for it can be done but might require some guessing or extension of the protocol. This was briefly discussed here: https://github.com/datopian/giftless/pull/68 but nothing was done to handle content-type, just content-disposition header. 

In any case if you want this supported, it should be reported as a new issue / feature request.

**Reply by ghost** [reply]
Here is how my original specs from gifteless.yaml for storage:

`TRANSFER_ADAPTERS:`
`  basic:`
`    factory: giftless.transfer.basic_streaming:factory`
`    options:`
`      storage_class: giftless.storage.google_cloud:GoogleCloudStorage`
`      storage_options:`
`        project_name: MyProject`
`        bucket_name: my_bucket`
`        account_key_file: account-key.jsnon`

Based on your reply I tried `giftless.transfer.basic_external:factory` and it seems to preserve the file extension correctly! 

I think it should be more clearly stated in the [documentation](https://giftless.datopian.com/en/latest/storage-backends.html)

**Comment by shevron** [comment]
Glad to know you have a solution, in general using the `external_storage` transfer adapter is recommended. 
Re documentation, it's a good idea to document it although I'd rather have it reported as a limitation / feature request in the issue tracker and not as a known attribute of the adapter, so if you wish I'd welcome the bug report.
